File: character_logic/happyrin.py
"""Happyrin battle and AI hooks."""

from __future__ import annotations

from typing import Any


CHARACTER_ID = "happyrin"
MADNESS = "광증"
MAX_MADNESS = 10
PASSIVE_AMOUNT = 7


def _madness_stacks(fighter: Any) -> int:
    status = fighter.statuses.get(MADNESS)
    return int(status.stacks) if status else 0


def _add_madness(battle: Any, fighter: Any, turns: int, stacks: int, source: str) -> None:
    battle.add_status(fighter, MADNESS, turns, stacks, source, stack=True, max_stacks=MAX_MADNESS)


def _madness_hit_cap(actor: Any, target: Any, replaced: bool) -> int:
    cap = _madness_stacks(actor)
    if replaced:
        cap += _madness_stacks(target)
    return max(0, cap)


def on_action_start_status(battle: Any, choice: Any) -> bool:
    actor = choice.actor
    original = choice.action
    status = actor.statuses.get(MADNESS)
    if not status or not original.is_active:
        return False

    chance = min(100, int(status.stacks) * 10)
    roll = battle.roll(MADNESS)
    print(f"광증 판정 {chance}% / 판정값 {roll:.2f}")
    if roll >= chance:
        return False

    options = []
    for slot in range(len(actor.data.get("skills", []))):
        action = actor.skill_by_key(battle.skill_key(actor.character_id, slot))
        if action and action.key != original.key:
            options.append(action)
    if not options:
        return False

    replacement = battle.rng.choice(options)
    choice.action = replacement
    choice.power = replacement.power
    choice.accuracy = replacement.accuracy
    choice.hit_count = 1
    choice.madness_replaced = True
    choice.madness_original_action_key = original.key
    battle.record.madness_decided[actor.side] = True
    _prepare_replacement_choice(actor, choice, replacement)
    print(f"광증으로 {original.name} 대신 {replacement.name}이 결정되었다.")
    return False


def _prepare_replacement_choice(actor: Any, choice: Any, action: Any) -> None:
    if actor.character_id == "gandrick":
        choice.selected_bullets = int(actor.counters.get("탄환", 0)) if action.is_skill("gandrick", 3) else None
    if actor.character_id == "balef" and action.is_active and action.is_attack and choice.prev_attack_active is None:
        history = actor.selected_attack_active_history
        choice.prev_attack_active = history[-1] if history else None


def apply_condition_effects(battle: Any, choice: Any) -> bool | None:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    if action.is_skill(CHARACTER_ID, 1):
        _add_madness(battle, actor, 2, 1, action.name)
    elif action.is_skill(CHARACTER_ID, 2):
        hit_cap = _madness_hit_cap(actor, target, choice.madness_replaced)
        if hit_cap < 1:
            return False
        choice.hit_count = battle.rng.randint(1, hit_cap)
        print(f"[연격] 광증 중첩 수 {hit_cap} 기준으로 {choice.hit_count}회로 결정되었다.")
    return None


def estimated_hit_count(actor: Any, action: Any, use_max: bool) -> float | None:
    if action.is_skill(CHARACTER_ID, 2):
        hit_cap = max(1, _madness_stacks(actor))
        return hit_cap if use_max else (1 + hit_cap) / 2
    return None


def attack_damage_multipliers(battle: Any, choice: Any) -> list[float]:
    if choice.action.is_skill(CHARACTER_ID, 1) and choice.madness_replaced:
        return [2.0]
    return []


def on_hit_after_defense(battle: Any, choice: Any, total_damage: int) -> None:
    actor = choice.actor
    target = battle.opponent(actor)
    action = choice.action
    if action.is_skill(CHARACTER_ID, 0):
        recipient = target if choice.madness_replaced else actor
        _add_madness(battle, recipient, 5, 1, action.name)
    elif action.is_skill(CHARACTER_ID, 1):
        stacks = _madness_stacks(actor)
        if choice.madness_replaced:
            stacks *= 2
        _add_madness(battle, target, 4, stacks, action.name)


def apply_non_attack_effects(battle: Any, choice: Any) -> bool:
    actor = choice.actor
    action = choice.action
    if not action.is_skill(CHARACTER_ID, 3):
        return False
    _add_madness(battle, actor, 4, 5, action.name)
    multiplier = 1.7 if choice.madness_replaced else 1.2
    for stat in ("atk", "def", "spd"):
        battle.add_stat_effect(actor, stat, multiplier, 4, action.name)
    return True


def on_turn_end(battle: Any, fighter: Any) -> None:
    if fighter.character_id != CHARACTER_ID:
        return
    opponent = battle.opponent(fighter)
    if battle.record.madness_decided.get(fighter.side):
        battle.heal(fighter, PASSIVE_AMOUNT, "복약 지도")
    if battle.record.madness_decided.get(opponent.side):
        battle.fixed_damage(opponent, PASSIVE_AMOUNT, "복약 지도")


def setup_value(battle: Any, actor: Any, target: Any, action: Any) -> float:
    own_madness = _madness_stacks(actor)
    target_madness = _madness_stacks(target)
    if action.is_skill(CHARACTER_ID, 3):
        incoming = battle.estimate_best_incoming_damage(target, actor)
        if own_madness < 5:
            value = 3500 if actor.hp > incoming * 1.15 else 950
            if action.mp <= actor.mp <= action.mp + 16:
                value += 520
            return value
        return 240
    if action.is_skill(CHARACTER_ID, 1):
        return max(180, (MAX_MADNESS - target_madness) * 70)
    if action.is_skill(CHARACTER_ID, 0) and own_madness < 3:
        return 220
    return 0.0


def _active_skill_actions(battle: Any, actor: Any) -> list[Any]:
    actions: list[Any] = []
    for slot in range(len(actor.data.get("skills", []))):
        action = actor.skill_by_key(battle.skill_key(actor.character_id, slot))
        if action is not None:
            actions.append(action)
    return actions


def _madness_result_value(
    battle: Any,
    actor: Any,
    target: Any,
    action: Any,
    replaced: bool,
    incoming: float,
) -> float:
    own_madness = _madness_stacks(actor)
    target_madness = _madness_stacks(target)
    hit_rate = battle.estimate_hit_rate(actor, target, action)
    damage = battle.estimate_action_damage(actor, target, action, use_max=False)

    if replaced and action.is_skill(CHARACTER_ID, 1):
        damage *= 2.0
    if replaced and action.is_skill(CHARACTER_ID, 2):
        one_hit = battle.calculate_estimated_damage(actor, target, action)
        hit_cap = _madness_hit_cap(actor, target, replaced=True)
        damage = one_hit * ((1 + hit_cap) / 2)

    expected_damage = damage * hit_rate
    value = expected_damage * 3.2
    if action.is_attack and expected_damage >= target.hp:
        value += 7200
    elif action.is_attack and damage >= target.hp:
        value += 4600 * hit_rate

    if action.is_skill(CHARACTER_ID, 0):
        if replaced:
            value += max(0, min(1, MAX_MADNESS - target_madness)) * 360 * hit_rate
        else:
            value += 180 if own_madness < 4 else -180
    elif action.is_skill(CHARACTER_ID, 1):
        own_after_condition = min(MAX_MADNESS, own_madness + 1)
        stacks = own_after_condition * (2 if replaced else 1)
        applied = min(MAX_MADNESS - target_madness, stacks)
        value += max(0, applied) * (145 + own_after_condition * 18) * hit_rate
        if own_madness >= 8:
            value -= 220
    elif action.is_skill(CHARACTER_ID, 2):
        hit_cap = _madness_hit_cap(actor, target, replaced)
        if hit_cap < 1:
            value -= 2400
        else:
            value += min(10, hit_cap) * 115 * hit_rate
        if replaced and target_madness >= 2:
            value += min(6, target_madness) * 140 * hit_rate
    elif action.is_skill(CHARACTER_ID, 3):
        multiplier = 1.7 if replaced else 1.2
        value += 950 if multiplier < 1.5 else 2550
        value += max(0, actor.max_hp - actor.hp) * (0.8 if multiplier < 1.5 else 1.25)
        if incoming >= actor.hp:
            value -= 650
        if own_madness >= 8:
            value -= 240

    if replaced:
        value += PASSIVE_AMOUNT * 42
    return value


def _madness_decoy_bonus(battle: Any, actor: Any, target: Any, action: Any, incoming: float) -> float:
    own_madness = _madness_stacks(actor)
    if own_madness <= 0 or not action.is_active:
        return 0.0

    proc = min(1.0, own_madness / 10)
    options = [candidate for candidate in _active_skill_actions(battle, actor) if candidate.key != action.key]
    if not options:
        return 0.0

    direct = _madness_result_value(battle, actor, target, action, replaced=False, incoming=incoming)
    replacement_values = [
        _madness_result_value(battle, actor, target, candidate, replaced=True, incoming=incoming)
        for candidate in options
    ]
    replacement_average = sum(replacement_values) / len(replacement_values)
    bonus = (replacement_average - direct) * proc

    if own_madness >= 6 and replacement_average > direct:
        best_replacement = max(replacement_values)
        bonus += max(0.0, best_replacement - replacement_average) * proc * 0.18
        bonus *= 1 + (own_madness - 5) * 0.22
        bonus = min(bonus, 1150 + own_madness * 70)
        pressure = incoming / max(1, actor.hp)
        hp_ratio = actor.hp / max(1, actor.max_hp)
        if pressure >= 1.0:
            bonus *= 0.2
        elif pressure >= 0.65 and hp_ratio < 0.55:
            bonus *= 0.4
        elif pressure >= 0.45 and hp_ratio < 0.45:
            bonus *= 0.65
    return bonus


def ai_score(
    battle: Any,
    actor: Any,
    target: Any,
    action: Any,
    expected_damage: float,
    hit_rate: float,
) -> float:
    value = 0.0
    own_madness = _madness_stacks(actor)
    target_madness = _madness_stacks(target)
    incoming = battle.estimate_best_incoming_damage(target, actor)

    value += _madness_decoy_bonus(battle, actor, target, action, incoming)

    if own_madness >= 6 and action.is_active:
        value -= (own_madness - 5) * 130
    if own_madness >= 7 and action.is_common_action("meditation"):
        value += 260

    if action.is_skill(CHARACTER_ID, 0):
        value += max(0, 4 - own_madness) * 120
        if own_madness == 0:
            if action.mp <= actor.mp < 37 and actor.mp + 10 < 37:
                value += 900
            elif actor.mp + 10 >= 37:
                value -= 260
        elif own_madness <= 2:
            value += 360
        if own_madness >= 6:
            value -= 180
    elif action.is_skill(CHARACTER_ID, 1):
        value += max(1, own_madness + 1) * 150 * hit_rate
        value += max(0, 6 - target_madness) * 90
    elif action.is_skill(CHARACTER_ID, 2):
        if expected_damage >= target.hp:
            value += 1800
        else:
            value += max(0, own_madness - 1) * 170 * hit_rate
    elif action.is_skill(CHARACTER_ID, 3):
        value += 620
        if own_madness < 5:
            value += 800
            if actor.hp > incoming * 1.15:
                value += 720
        if incoming >= actor.hp:
            value -= 500
    return value


def would_condition_fail(battle: Any, actor: Any, target: Any, action: Any) -> bool | None:
    if action.is_skill(CHARACTER_ID, 2):
        return _madness_stacks(actor) < 1
    return None

File: web/battle-engine/character-logic/happyrin.js
"use strict";

const CHARACTER_ID = "happyrin";
const MADNESS = "광증";
const MAX_MADNESS = 10;

function madnessStacks(fighter) {
  return Number(fighter.statuses[MADNESS]?.stacks || 0);
}

function addMadness(battle, fighter, turns, stacks, source) {
  battle.addStatus(fighter, MADNESS, turns, stacks, source, true, MAX_MADNESS);
}

function randomIntInclusive(battle, low, high) {
  return low + battle.rng.range(high - low + 1);
}

function activeSkillActions(battle, actor) {
  return battle.availableActions(actor).filter((action) => action.isActive);
}

function madnessResultValue(battle, actor, target, action, replaced, incoming) {
  const ownMadness = madnessStacks(actor);
  const targetMadness = madnessStacks(target);
  const hitRate = battle.estimateHitRate(actor, target, action) / 100;
  let damage = battle.estimateActionDamage(actor, target, action, false);
  if (replaced && action.isSkill(CHARACTER_ID, 1)) damage *= 2;
  if (replaced && action.isSkill(CHARACTER_ID, 2)) {
    const oneHit = battle.calculateEstimatedDamage(actor, target, action);
    const hitCap = madnessStacks(actor) + madnessStacks(target);
    damage = oneHit * ((1 + hitCap) / 2);
  }
  const expectedDamage = damage * hitRate;
  let value = expectedDamage * 3.2;
  if (action.isAttack && expectedDamage >= target.hp) value += 7200;
  else if (action.isAttack && damage >= target.hp) value += 4600 * hitRate;

  if (action.isSkill(CHARACTER_ID, 0)) {
    if (replaced) value += Math.max(0, Math.min(1, MAX_MADNESS - targetMadness)) * 360 * hitRate;
    else value += ownMadness < 4 ? 180 : -180;
  } else if (action.isSkill(CHARACTER_ID, 1)) {
    const ownAfterCondition = Math.min(MAX_MADNESS, ownMadness + 1);
    const stacks = ownAfterCondition * (replaced ? 2 : 1);
    const applied = Math.min(MAX_MADNESS - targetMadness, stacks);
    value += Math.max(0, applied) * (145 + ownAfterCondition * 18) * hitRate;
    if (ownMadness >= 8) value -= 220;
  } else if (action.isSkill(CHARACTER_ID, 2)) {
    const hitCap = ownMadness + (replaced ? targetMadness : 0);
    if (hitCap < 1) value -= 2400;
    else value += Math.min(10, hitCap) * 115 * hitRate;
    if (replaced && targetMadness >= 2) value += Math.min(6, targetMadness) * 140 * hitRate;
  } else if (action.isSkill(CHARACTER_ID, 3)) {
    const multiplier = replaced ? 1.7 : 1.2;
    value += multiplier < 1.5 ? 950 : 2550;
    value += Math.max(0, actor.maxHp - actor.hp) * (multiplier < 1.5 ? 0.8 : 1.25);
    if (incoming >= actor.hp) value -= 650;
    if (ownMadness >= 8) value -= 240;
  }
  if (replaced) value += 7 * 42;
  return value;
}

function madnessDecoyBonus(battle, actor, target, action, incoming) {
  const ownMadness = madnessStacks(actor);
  if (ownMadness <= 0 || !action.isActive) return 0;
  const proc = Math.min(1, ownMadness / 10);
  const options = activeSkillActions(battle, actor).filter((candidate) => candidate.key !== action.key);
  if (!options.length) return 0;
  const direct = madnessResultValue(battle, actor, target, action, false, incoming);
  const replacementValues = options.map((candidate) => madnessResultValue(battle, actor, target, candidate, true, incoming));
  const replacementAverage = replacementValues.reduce((sum, item) => sum + item, 0) / replacementValues.length;
  let bonus = (replacementAverage - direct) * proc;
  if (ownMadness >= 6 && replacementAverage > direct) {
    const bestReplacement = Math.max(...replacementValues);
    bonus += Math.max(0, bestReplacement - replacementAverage) * proc * 0.18;
    bonus *= 1 + (ownMadness - 5) * 0.22;
    bonus = Math.min(bonus, 1150 + ownMadness * 70);
    const pressure = incoming / Math.max(1, actor.hp);
    const hpRatio = actor.hp / Math.max(1, actor.maxHp);
    if (pressure >= 1) bonus *= 0.2;
    else if (pressure >= 0.65 && hpRatio < 0.55) bonus *= 0.4;
    else if (pressure >= 0.45 && hpRatio < 0.45) bonus *= 0.65;
  }
  return bonus;
}

function rawDamage(expectedDamage, hitRate) { return hitRate > 0 ? expectedDamage / hitRate : 0; }

module.exports = {
  onActionStartStatus(battle, choice) {
    const actor = choice.actor;
    const original = choice.action;
    if (choice.actionReplacementLocked) return false;
    const status = actor.statuses[MADNESS];
    if (!status || !original.isActive) return false;
    const chance = Math.min(100, Number(status.stacks) * 10);
    const roll = battle.roll(MADNESS);
    battle.logs.push(`광증 판정 ${chance}% / 판정값 ${roll.toFixed(2)}`);
    if (roll >= chance) return false;
    const options = activeSkillActions(battle, actor).filter((action) => action.key !== original.key);
    if (!options.length) return false;
    const replacement = battle.rng.choice(options);
    choice.action = replacement;
    choice.power = replacement.power;
    choice.accuracy = replacement.accuracy;
    choice.hitCount = 1;
    choice.madnessReplaced = true;
    choice.madnessOriginalActionKey = original.key;
    if (replacement.characterId === "gandrick") {
      choice.selectedBullets = replacement.isSkill("gandrick", 3) ? Number(actor.counters["탄환"] || 0) : null;
    }
    if (replacement.characterId === "balef" && replacement.isActive && replacement.isAttack && choice.prevAttackActive == null) {
      const selected = actor.selectedAttackActiveHistory || [];
      choice.prevAttackActive = selected.length ? selected[selected.length - 1] : null;
    }
    if (!battle.record.madnessDecided) battle.record.madnessDecided = {};
    battle.record.madnessDecided[actor.side] = true;
    battle.logs.push(`광증으로 ${original.name} 대신 ${replacement.name}이 결정되었다.`);
    return false;
  },

  applyConditionEffects(battle, choice) {
    const actor = choice.actor;
    const target = battle.opponent(actor);
    if (choice.action.isSkill(CHARACTER_ID, 1)) {
      addMadness(battle, actor, 2, 1, choice.action.name);
    } else if (choice.action.isSkill(CHARACTER_ID, 2)) {
      let cap = madnessStacks(actor);
      if (choice.madnessReplaced) cap += madnessStacks(target);
      if (cap < 1) return false;
      choice.hitCount = randomIntInclusive(battle, 1, cap);
      battle.logs.push(`[연격] 광증 중첩 수 ${cap} 기준으로 ${choice.hitCount}회로 결정되었다.`);
    }
    return true;
  },

  attackDamageMultipliers(_battle, choice) {
    return choice.action.isSkill(CHARACTER_ID, 1) && choice.madnessReplaced ? [2] : [];
  },

  onHitAfterDefenseAsActor(battle, choice) {
    const actor = choice.actor;
    const target = battle.opponent(actor);
    if (choice.action.isSkill(CHARACTER_ID, 0)) {
      addMadness(battle, choice.madnessReplaced ? target : actor, 5, 1, choice.action.name);
    } else if (choice.action.isSkill(CHARACTER_ID, 1)) {
      let stacks = madnessStacks(actor);
      if (choice.madnessReplaced) stacks *= 2;
      addMadness(battle, target, 4, stacks, choice.action.name);
    }
  },

  applyNonAttackEffects(battle, choice) {
    if (!choice.action.isSkill(CHARACTER_ID, 3)) return false;
    addMadness(battle, choice.actor, 4, 5, choice.action.name);
    const multiplier = choice.madnessReplaced ? 1.7 : 1.2;
    for (const stat of ["atk", "def", "spd"]) battle.addStatEffect(choice.actor, stat, multiplier, 4, choice.action.name);
    return true;
  },

  onTurnEnd(battle, fighter) {
    if ((battle.activeCharacterId?.(fighter) || fighter.characterId) !== CHARACTER_ID) return;
    const decided = battle.record.madnessDecided || {};
    const opponent = battle.opponent(fighter);
    if (decided[fighter.side]) battle.heal(fighter, 7, "복약 지도");
    if (decided[opponent.side]) battle.fixedDamage(opponent, 7, "복약 지도", fighter);
  },

  setupValue(battle, actor, target, action) {
    const ownMadness = madnessStacks(actor);
    const targetMadness = madnessStacks(target);
    if (action.isSkill(CHARACTER_ID, 3)) {
      const incoming = battle.estimateBestIncomingDamage(target, actor);
      if (ownMadness < 5) {
        let value = actor.hp > incoming * 1.15 ? 3500 : 950;
        if (action.mp <= actor.mp && actor.mp <= action.mp + 16) value += 520;
        return value;
      }
      return 240;
    }
    if (action.isSkill(CHARACTER_ID, 1)) return Math.max(180, (MAX_MADNESS - targetMadness) * 70);
    if (action.isSkill(CHARACTER_ID, 0) && ownMadness < 3) return 220;
    return 0;
  },

  aiScore(battle, actor, target, action, expectedDamage, hitRate) {
    let value = 0;
    const damage = rawDamage(expectedDamage, hitRate);
    const ownMadness = madnessStacks(actor);
    const targetMadness = madnessStacks(target);
    const incoming = battle.estimateBestIncomingDamage(target, actor);
    value += madnessDecoyBonus(battle, actor, target, action, incoming);
    if (ownMadness >= 6 && action.isActive) value -= (ownMadness - 5) * 130;
    if (ownMadness >= 7 && action.isCommonAction("meditation")) value += 260;

    if (action.isSkill(CHARACTER_ID, 0)) {
      value += Math.max(0, 4 - ownMadness) * 120;
      if (ownMadness === 0) {
        const turnMp = battle.turnEndMpRecovery(actor);
        if (action.mp <= actor.mp && actor.mp < 37 && actor.mp + turnMp < 37) value += 900;
        else if (actor.mp + turnMp >= 37) value -= 260;
      } else if (ownMadness <= 2) value += 360;
      if (ownMadness >= 6) value -= 180;
    } else if (action.isSkill(CHARACTER_ID, 1)) {
      value += Math.max(1, ownMadness + 1) * 150 * hitRate;
      value += Math.max(0, 6 - targetMadness) * 90;
    } else if (action.isSkill(CHARACTER_ID, 2)) {
      if (damage >= target.hp) value += 1800;
      else value += Math.max(0, ownMadness - 1) * 170 * hitRate;
    } else if (action.isSkill(CHARACTER_ID, 3)) {
      value += 620;
      if (ownMadness < 5) {
        value += 800;
        if (actor.hp > incoming * 1.15) value += 720;
      }
      if (incoming >= actor.hp) value -= 500;
    }
    return value;
  },

  estimatedHitCount(actor, action, useMax) {
    if (!action.isSkill(CHARACTER_ID, 2)) return null;
    const cap = Math.max(1, madnessStacks(actor));
    return useMax ? cap : (1 + cap) / 2;
  },

  wouldConditionFail(_battle, actor, _target, action) {
    return action.isSkill(CHARACTER_ID, 2) ? madnessStacks(actor) < 1 : null;
  },
};
